mapiproxy/servers/default/emsmdb/emsmdbp_object.c

```diff
@@ -3138,6 +3138,150 @@ _PUBLIC_ struct emsmdbp_object *emsmdbp_object_ftcontext_init(TALLOC_CTX *mem_ct
+	} else {
+		/* This shouldn't happen, but just in case */
+		xml = "";
+		OPENCHANGE_RETVAL_IF(!xml, MAPI_E_NOT_ENOUGH_MEMORY, local_mem_ctx);
+	}
+
+	*target_recipients_attr = talloc_strdup(mem_ctx, xml);
+	OPENCHANGE_RETVAL_IF(!*target_recipients_attr, MAPI_E_NOT_ENOUGH_MEMORY, local_mem_ctx);
+
+	talloc_free(local_mem_ctx);
+
+	return MAPI_E_SUCCESS;
+}
+
+/**
+   \details Store the Providers node into providers_node out parameter.
+
+   \param emsmdbp_ctx pointer to the emsmdb provider cotnext
+   \param sharing_object the sharing message object to get the XML data from
+   \param remote_uid string with the folder id to share, NULL when no required
+   \param remote_store_uid string with the mailbox id where the folder belongs to, NULL when non-required
+   \param target_recipients the string where the recipients' email addresses separated by semi colons are stored
+   \param [out] providers_node pointer to string where the Providers XML node is stored
+   \param mem_ctx pointer to memory context where the providers_node is created
+
+   \note On success providers_node string is allocated in mem_ctx. It is up to the developer
+   to free that memory context when it is not needed anymore.
+
+   \return MAPI_E_SUCCESS on success, a mapi error code otherwise
+ */
+static enum MAPISTATUS emsmdbp_object_sharing_metadata_providers_node(struct emsmdbp_context *emsmdbp_ctx, struct emsmdbp_object *sharing_object, char *remote_uid, char *remote_store_uid, char *target_recipients, char **providers_node, TALLOC_CTX *mem_ctx)
+{
+	char	       *xml;
+	TALLOC_CTX     *local_mem_ctx;
+
+	/* Sanity checks */
+	OPENCHANGE_RETVAL_IF(!emsmdbp_ctx, MAPI_E_INVALID_PARAMETER, NULL);
+	OPENCHANGE_RETVAL_IF(!sharing_object, MAPI_E_INVALID_PARAMETER, NULL);
+	OPENCHANGE_RETVAL_IF(!mem_ctx, MAPI_E_INVALID_PARAMETER, NULL);
+
+	local_mem_ctx = talloc_named(NULL, 0, "sharing_metadata_providers_node");
+	OPENCHANGE_RETVAL_IF(!local_mem_ctx, MAPI_E_NOT_ENOUGH_MEMORY, NULL);
+
+	xml = talloc_asprintf(local_mem_ctx, "<Providers><Provider Type=\"ms-exchange-internal\" %s >", target_recipients);
+	OPENCHANGE_RETVAL_IF(!xml, MAPI_E_NOT_ENOUGH_MEMORY, local_mem_ctx);
+
+	if (remote_uid && remote_store_uid) {
+		xml = talloc_asprintf_append(xml,
+					     "<FolderId xmlns=\"http://schemas.microsoft.com/exchange/sharing/2008\">%s</FolderId>",
+					     remote_uid);  /* PidLidSharingRemoteUid */
+		OPENCHANGE_RETVAL_IF(!xml, MAPI_E_NOT_ENOUGH_MEMORY, local_mem_ctx);
+
+		xml = talloc_asprintf_append(xml,
+					     "<MailboxId xmlns=\"http://schemas.microsoft.com/exchange/sharing/2008\">%s</MailboxId>",
+					     remote_store_uid);	 /* PidLidSharingRemoteStoreUid */
+		OPENCHANGE_RETVAL_IF(!xml, MAPI_E_NOT_ENOUGH_MEMORY, local_mem_ctx);
+	}
+
+	xml = talloc_asprintf_append(xml, "</Provider></Providers>");
+	OPENCHANGE_RETVAL_IF(!xml, MAPI_E_NOT_ENOUGH_MEMORY, local_mem_ctx);
+
+	*providers_node = talloc_strdup(mem_ctx, xml);
```

Comment:
either check `providers_node != NULL` before dereference it here or at beginning of the function with the rest of sanity checks.